Fix a getauxval comment and expand the Krait idiv workaround

rprichard opened the pull request:
The next NDK (r24) raises the minimum supported Android API from 16 to 19. That's new enough that all supported targets with NDK r24 have the getauxval function. i.e. Maybe google/cpu_features can drop support for dlopen/dlsym of getauxval at some point.

I noticed the lack of idiv support on my 2013 Nexus 7 device, because I used that device to run the NDK's tests. One of the tests verifies that the idiv support reported by the NDK cpu-features library matches what the device actually supports. Both the NDK's cpu-features and this cpu_features library only recognize the 0x510006f2 and 0x510006f3 CPU IDs for this workaround.

See https://github.com/android/ndk/issues/1605.

rprichard commented:
I'm wondering if the check should be more conservative. Maybe this instead:

```
  // Some Qualcomm Krait kernels forget to report IDIV support.
  // https://github.com/torvalds/linux/commit/120ecfafabec382c4feb79ff159ef42a39b6d33b
  if (info->implementer == 0x51 && info->architecture == 7 &&
      (info->part == 0x4d || info->part == 0x6f)) {
    info->features.idiva = true;
    info->features.idivt = true;
  }
```

I can't find good documentation for the (0x4d or 0x6f), but I see a few places:
 - Wikipedia: https://en.wikipedia.org/wiki/Comparison_of_ARMv7-A_cores
 - pytorch also looks for 0x4d or 0x6f: https://github.com/pytorch/cpuinfo/commit/ac57632afc7a7c44f75544e89859426a37d6d692#diff-02fec0405ceaca10fb9a8ada5cd3ffa9090948a4993f2461ec7c8694e9ff2734R118-R125
 - 0x4d devices on https://crbug.com/341598:
   - kernel reports idiv: [comment 33](https://crbug.com/341598#c33)
   - kernel reports no idiv: [comment 39](https://crbug.com/341598#c39)
   - kernel reports no idiv: [comment 43](https://crbug.com/341598#c43)

It looks like the only Krait part numbers are 0x4d (dual-core) and 0x6f (quad-core).

toor1245 commented:
> I'm wondering if the check should be more conservative. Maybe this instead:
> 
> ```c
>   // Some Qualcomm Krait kernels forget to report IDIV support.
>   // https://github.com/torvalds/linux/commit/120ecfafabec382c4feb79ff159ef42a39b6d33b
>   if (info->implementer == 0x51 && info->architecture == 7 &&
>       (info->part == 0x4d || info->part == 0x6f)) {
>     info->features.idiva = true;
>     info->features.idivt = true;
>   }
> ```
> 
> I can't find good documentation for the (0x4d or 0x6f), but I see a few places:
> 
> * Wikipedia: https://en.wikipedia.org/wiki/Comparison_of_ARMv7-A_cores
> * pytorch also looks for 0x4d or 0x6f: [pytorch/cpuinfo@ac57632#diff-02fec0405ceaca10fb9a8ada5cd3ffa9090948a4993f2461ec7c8694e9ff2734R118-R125](https://github.com/pytorch/cpuinfo/commit/ac57632afc7a7c44f75544e89859426a37d6d692#diff-02fec0405ceaca10fb9a8ada5cd3ffa9090948a4993f2461ec7c8694e9ff2734R118-R125)
> * 0x4d devices on https://crbug.com/341598:
>   
>   * kernel reports idiv: [comment 33](https://crbug.com/341598#c33)
>   * kernel reports no idiv: [comment 39](https://crbug.com/341598#c39)
>   * kernel reports no idiv: [comment 43](https://crbug.com/341598#c43)
> 
> It looks like the only Krait part numbers are 0x4d (dual-core) and 0x6f (quad-core).

I think it would be good if we switched to this implementation. I also searched everywhere but didn't find anything new, since VFPv4 is an optional implementation, I would rather not specify the IDIV flag for everyone with part number in the range 0x40 to 0x7F, but only specify 0x4d and 0x6f, what do you think?

rprichard commented:
> I think it would be good if we switched to this implementation. I also searched everywhere but didn't find anything new, since VFPv4 is an optional extension, I would rather not specify the IDIV flag for everyone with part number in the range 0x40 to 0x7F, but only specify 0x4d and 0x6f, what do you think?

Yeah, that seems better to me. I updated the patch to do that.

rprichard commented:
Is this change OK to merge now?

toor1245 commented:
@rprichard I have no objection to your changes, @gchatelet could you take a look?
